controllers/datamovement_controller.go

```diff
@@ -414,29 +459,97 @@ func getCmdAndArgs(cmCommand string, cmNumProcesses int, progressInterval int, h
 		cmd = cmdList[0]
 		args = cmdList[1:]
 	} else {
-		numProcesses := len(hosts)
+		cmd = "mpirun"
+
+		// Required mpirun options
+		args = []string{
+			"--allow-run-as-root", // required to be able to set dcp uid/gid
+			"--hostfile", hostfilePath,
+		}
 
-		// If data movement is on the rabbit namespace, use the config map value
+		// If we're doing data movement on the Rabbit, we can specify how many (local) processors to
+		// contribute towards dcp. However, if this is a lustre2lustre, -np should still be the
+		// number of rabbits (len(hosts)).
+		np := numHosts
 		if dm.Namespace == os.Getenv("NNF_NODE_NAME") {
-			numProcesses = cmNumProcesses
+			np = numProcesses
 		}
 
-		cmd = "mpirun"
-		args = []string{
-			"--allow-run-as-root", // required to be able to set dcp uid/gid
-			"-np", fmt.Sprintf("%d", numProcesses),
-			"--host", strings.Join(hosts, ","),
+		// mpirun -np is optional when using hostfiles
+		if np > -1 {
+			numProcessesArgs := []string{
+				"-np", fmt.Sprintf("%d", np),
+			}
+			args = append(args, numProcessesArgs...)
+		}
+
+		// Additional mpi options
+		if len(mpiOpts) > 0 {
+			opts := strings.Split(mpiOpts, " ")
+			args = append(args, opts...)
+		}
+
+		// Required dcp options
+		dcpArgs := []string{
 			"dcp",
 			"--progress", fmt.Sprintf("%d", progressInterval),
 			"--uid", fmt.Sprintf("%d", dm.Spec.UserId),
 			"--gid", fmt.Sprintf("%d", dm.Spec.GroupId),
-			dm.Spec.Source.Path, dm.Spec.Destination.Path,
 		}
+
+		// Additional dcp options
+		if len(dcpOpts) > 0 {
+			opts := strings.Split(dcpOpts, " ")
+			dcpArgs = append(dcpArgs, opts...)
+		}
+
+		// dcp source/path arguments
+		dcpArgs = append(dcpArgs, dm.Spec.Source.Path)
+		dcpArgs = append(dcpArgs, dm.Spec.Destination.Path)
+
+		// Add dcp command to mpirun command
+		args = append(args, dcpArgs...)
 	}
 
 	return cmd, args
 }
 
+// Create an MPI Hostfile given a list of hosts, slots, and maxSlots. A temporary directory is
+// created based on the DM Name. The hostfile is created inside of this directory.
+func createMpiHostfile(dmName string, hosts []string, slots, maxSlots int) (string, error) {
```

Comment:
If the directory is unique, why is it needed? Just name the host file what you are currently using for the directory name

Reply:
I thought it could be useful in the future if we needed other supporting files/configuration for data movement. Since we've opened extra options to dcp and mpirun, both have some options that take in files. I figured a directory would be a good place to stash these files if we ever got there.